### ISSUE TITLE
UX improvements for POIs

### ISSUE THREAD

**Issue by pingurus** [Maproom/qmapshack#313]
**Important Notice:**

This is for developers only. As a user use the `bug report` or `Everything that's not a bug` template.

**The parent ticket**

_add ticketnumber behind the #_

QMS-#294

**What do you plan to do**
Improve the user experience by feedback provided.

**The result will be**

(Remarks by @wthaemelt in #309 , comments in italics by @pingurus)
* [ ] Show the icon for the POI category in the POI selection window next to the category name. The meaning of an icon in the map window is not always obvious.
* [x] _(probably wontfix)_ In the POI database names of POI are provided in different languages. Which one is selected for use in QMS? QMS language? System language? _Whatever the language in the 'name=' tag is. Should be english._
* [x] Add the usual tooltip info when hovering over a POI on the map.
* [x] What happens, if some POI have the same location? _Right now nothing. Also note that some POIs fall in multiple categories_ There a nice solution for waypoints in this situation.
* [ ] A serious issue seems to be the time necessary for loading POIs (even when outside the visible area of the map). The GUI is blocked. A "busy" cursor appears only after some further clicks. This issue is similar to issue #310 - the user doesn't get sufficient information that he has to wait and how long he has to wait. If there is the default cursor, then there is no information at all that QMS is still busy. Waiting several minutes (and this happened even with a small selection of POI) for the next GUI interaction is hardly accepted by the user. _Also related to #307_
* [ ] _(probably wontfix)_ Would it be possible to add the number of POIs per category to the POI selection list? Then the user would get an early information about the workload he is going to select. _We would first have to count them, which is exactly the same workload for the SQL Query as showing them_
* [x] The blue circle appearing when moving the mouse on a POI is rarely visible on dark icons. If icons are very close one to the other, then it is difficult to understand which icon is hit. Proposal: make the blue circle bigger than the icon.

**Comment by kiozen** [comment]
my 2 cents:

* the count function of the database should be more efficient than doing it inside our code:

**Comment by wthaem** [comment]
SQL proposal:

```
SELECT COUNT(P.ID) AS CNT, C.CATEGORY
FROM [POI_Data] P JOIN POI_CATEGORY_MAP C
ON P.ID = C.ID
GROUP BY C.CATEGORY
ORDER BY C.CATEGORY
.```

Result with 234598 POI and 331 categories when using my SQLite browser:

    304 Reihen innerhalb von 1205ms zurückgegeben

POI file size 55.4 MB. 

Duration mentioned doesn't include loading of file!

**Comment by wthaem** [comment]
The poi_index table in a POI file is a virtual table to which 3 tables poi_index_* are attached. Using SQLite with rtree extension implemented these tables improve the performance of queries with a geographic bounding box considerably.

Rtree support in SQLite is switched off by default and special precaution must be taken to activate it during compilation. 

Can the discussed relatively bad performance of POI handling result from missing rtree support in SQLite? I guess, my QMS installation doesn't have rtree support.

I tested a few queries with the latest version of "DB browser for SQLite" which supports rtree. No problems with performance.

Can this become a future requirement for QMS compilation/distribution?

**Comment by pingurus** [comment]
> Rtree support in SQLite is switched off by default and special precaution must be taken to activate it during compilation.
>
> Can the discussed relatively bad performance of POI handling result from missing rtree support in SQLite? I guess, my QMS installation doesn't have rtree support.

It seems as it if possible to enable RTree support on Qt, however this is just a bycatch from a configuration displayed in a bug report: https://bugreports.qt.io/browse/QTBUG-63498

**Comment by wthaem** [comment]
Thanks for the hint.

Which SQLite module is used when working with POIs? Qt or the genuine SQLite? How to find out in a given installation if the `SQLITE_ENABLE_RTREE` switch was used when compiling the module? From what I've read performance seems to be very different in our case between "classical" SQL and SQL using rtree.

**Comment by wthaem** [comment]
Please have a look at

[http://sqlite.1065341.n5.nabble.com/RTree-query-td41279.html](http://sqlite.1065341.n5.nabble.com/RTree-query-td41279.html)

and compare with the SQL query structure in CPoiPOI,cpp.

Would the performance be better when in a first query you get quickly the necessary  `id` values from the index table (using rtree) and then do in a second query the rest of the work with these ids (without joining tables) as done in the linked article?

**Comment by pingurus** [comment]
> Please have a look at
> 
> http://sqlite.1065341.n5.nabble.com/RTree-query-td41279.html
> 
> and compare with the SQL query structure in CPoiPOI,cpp.
> 
> Would the performance be better when in a first query you get quickly the necessary `id` values from the index table (using rtree) and then do in a second query the rest of the work with these ids (without joining tables) as done in the linked article?

Thanks, that's worth a try.

**Comment by pingurus** [comment]
> Please have a look at
> 
> http://sqlite.1065341.n5.nabble.com/RTree-query-td41279.html
> 
> and compare with the SQL query structure in CPoiPOI,cpp.
> 
> Would the performance be better when in a first query you get quickly the necessary `id` values from the index table (using rtree) and then do in a second query the rest of the work with these ids (without joining tables) as done in the linked article?

I just compared the query that is currently used in code: 
```
SELECT main.poi_index.maxLat, main.poi_index.maxLon, main.poi_index.minLat, main.poi_index.minLon, main.poi_data.data
FROM main.poi_index
INNER JOIN main.poi_data ON main.poi_index.id=main.poi_data.id
WHERE main.poi_index.maxLat<49.3
AND main.poi_index.minLat>49.2
AND main.poi_index.maxLon<8.1
AND main.poi_index.minLon>8
AND main.poi_data.data LIKE '%tourism=camp_site%'
```
To a query styled like on the site you linked to:
```
SELECT main.poi_data.data
FROM main.poi_data
WHERE id IN
(
	SELECT main.poi_index.id 
	FROM main.poi_index
	WHERE main.poi_index.maxLat<49.3
	AND main.poi_index.minLat>49.2
	AND main.poi_index.maxLon<8.1
	AND main.poi_index.minLon>8
)
```
The second query does a filtering step less and already takes double the time in the DB Browser for SQLite compared to the first one (~25ms vs ~50ms), so I guess that either our use-case is somehow different or there have been changes in the last 12 years to the internal workings of SQLite.

**Comment by wthaem** [comment]
Timing seems to be a difficult issue:

I took Germany.poi (the big one with ~4000000 POIs), used DB Browser for SQLite, version 3.12.0, and got:

```
Ausführung wurde ohne Fehler beendet.
Ergebnis: 1 Zeilen in 60ms zurückgegeben
In Zeile 1:
SELECT main.poi_index.maxLat, main.poi_index.maxLon, main.poi_index.minLat, main.poi_index.minLon, main.poi_data.data
FROM main.poi_index
INNER JOIN main.poi_data ON main.poi_index.id=main.poi_data.id
WHERE main.poi_index.maxLat<49.3
AND main.poi_index.minLat>49.2
AND main.poi_index.maxLon<8.1
AND main.poi_index.minLon>8
AND main.poi_data.data LIKE '%tourism=camp_site%'



Ausführung wurde ohne Fehler beendet.
Ergebnis: 1 Zeilen in 25ms zurückgegeben
In Zeile 1:
SELECT main.poi_index.maxLat, main.poi_index.maxLon, main.poi_index.minLat, main.poi_index.minLon, main.poi_data.data
FROM main.poi_data, main.poi_index
WHERE main.poi_data.id IN
(
	SELECT main.poi_index.id 
	FROM main.poi_index
	WHERE main.poi_index.maxLat<49.3
	AND main.poi_index.minLat>49.2
	AND main.poi_index.maxLon<8.1
	AND main.poi_index.minLon>8
)
and main.poi_data.data LIKE '%tourism=camp_site%' <========== find first id and after that continue with reduced set
and main.poi_data.id = main.poi_index.id   
```

Just the opposite result. But comparisons of duration for just 1 test are not very reliable!

But compare with

```
Ausführung wurde ohne Fehler beendet.
Ergebnis: 1 Zeilen in 10380ms zurückgegeben
In Zeile 1:
SELECT main.poi_data.data
FROM main.poi_data
WHERE id IN
(
	SELECT main.poi_index.id 
	FROM main.poi_index
	WHERE main.poi_index.maxLat<49.3
	AND main.poi_index.minLat>49.2
	AND main.poi_index.maxLon<8.1
	AND main.poi_index.minLon>8
    and main.poi_data.data LIKE '%tourism=camp_site%'    
)
```

or even 

```
Ergebnis: 1 Zeilen in 49345ms zurückgegeben
In Zeile 1:
SELECT main.poi_index.maxLat, main.poi_index.maxLon, main.poi_index.minLat, main.poi_index.minLon, main.poi_data.data
FROM main.poi_data, main.poi_index
WHERE main.poi_data.id IN
(
	SELECT main.poi_index.id 
	FROM main.poi_index
	WHERE main.poi_index.maxLat<49.3
	AND main.poi_index.minLat>49.2
	AND main.poi_index.maxLon<8.1
	AND main.poi_index.minLon>8
	and main.poi_data.data LIKE '%tourism=camp_site%'
)
and main.poi_data.id = main.poi_index.id
```

**Summary:**

* Performance depends heavily on the way SQL queries are formulated and executed.
* It is likely that in the first 2 cases the rtree feature was used efficiently and ensured short query times (my DB browser supports rtree!).
* Recommendation: In a first query step do nothing else then find the `id` of the POIs in the considered area. Only in a second query step reduce the number of POIs further by applying more search criteria. My last 2 queries demonstrate the disastrous effect when not doing so.
* The QMS POI code is using the QtSql engine with QSQLITE. If I'm not mistaken, the necessary SQL driver isn't anymore distributed with newer Qt versions. You have to compile it before you can use SQLite with QtSql (I did this for Windows without any problems). QMS packages must take care of this requirement (I don't know if QtSql/SQLite is used somewhere else in QMS).
* Additionally, QtSql must support the rtree extension. Otherwise, performance of POI handling will be very weak (I used PyQt5 for a test - the QtSql version used there had this rtree support). I'm not sure if this might be an additional requirement for QMS packages.

**Comment by pingurus** [comment]
> I don't know if QtSql/SQLite is used somewhere else in QMS

If I'm not mistaken, the database files are SQLite files, if they are local.

**Comment by kiozen** [comment]
Workspace and local database is all SQLite.

We have no real influence on how the sqldriver is compiled. 

I can't tell if SQLite is distributed with Qt. I know there are license issues with MySql. But @huirad can tell for sure.

On my Linux system rtree seems to be compiled in. Simple test:

```
~> sqlite3 test.db
SQLite version 3.28.0 2019-04-16 19:49:53
Enter ".help" for usage hints.
sqlite> create virtual table xxx using rtree("x", "y", "z");
sqlite> 
```

**Comment by huirad** [comment]
> I can't tell if SQLite is distributed with Qt. I know there are license issues with MySql. But @huirad can tell for sure.

This is an interesting question ;-)
What I can tell you: as long as the Windows Installer is built as described in https://github.com/Maproom/qmapshack/wiki/BuildWindowsVisualStudio, the sqlite3.dll is simply copied from the binary sqlite distribution.
That sqlite3.dll is maninly provided to support some features of PROJ and GDAL.

Whether Qt contains an embedded (maybe statically linked) version of sqlite I don't know.
But you could try and see what happens when you rename the sqlite3.dll in the QMS installation directory. 
If your database still works, then probably a statically linked sqlite3 is used.

**Comment by wthaem** [comment]
There seems to be a misunderstanding. What I had in mind is the `qsqlpsql.dll` file located for Windows in the `sqldrivers` directory.

Compare:

[https://doc.qt.io/qt-5/sql-driver.html#how-to-build-the-qmysql-plugin-on-windows](https://doc.qt.io/qt-5/sql-driver.html#how-to-build-the-qmysql-plugin-on-windows)

**Comment by pingurus** [comment]
@wthaem This if fixed, right?

> A serious issue seems to be the time necessary for loading POIs (even when outside the visible area of the map). The GUI is blocked. A "busy" cursor appears only after some further clicks. This issue is similar to issue #310 - the user doesn't get sufficient information that he has to wait and how long he has to wait. If there is the default cursor, then there is no information at all that QMS is still busy. Waiting several minutes (and this happened even with a small selection of POI) for the next GUI interaction is hardly accepted by the user.

**Comment by wthaem** [comment]
I did not have a chance to have a look at the latest commits. My hope is, that the bounding box use improves the situation very much. Thus, for the moment being, I would say "fixed".

But what about the cursor size and color many times mentioned?

**Comment by pingurus** [comment]
> But what about the cursor size and color many times mentioned?

If I remember correctly @kiozen said that there was no larger one, but I don't know whether we could maybe make it another color.

**Comment by wthaem** [comment]
Move a map so that it takes a while to redraw it (without POIs), then you see a bigger one!

**Comment by pingurus** [comment]
Sure, but then those two would lie on top of each other, so one gets no feedback what is still loading, only that something is loading.
I think that that also would be a valid solution, but I'm also fine with the current solution.
@kiozen what is your opinion on this?
